File: steganography/programs/encode.cpp
#include <vector>
#include "splashkit.h"

using std::to_string;

string BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

string base64_encode(const string &input)
{
    string encoded_string;
    int value = 0, bits = -6;
    const unsigned int base64_mask = 0x3F;

    for (unsigned char character : input)
    {
        value = (value << 8) + character;
        bits += 8;

        while (bits >= 0)
        {
            encoded_string.push_back(BASE64_CHARS[(value >> bits) & base64_mask]);
            bits -= 6;
        }
    }

    if (bits > -6)
    {
        encoded_string.push_back(BASE64_CHARS[((value << 8) >> (bits + 8)) & base64_mask]);
    }

    while (encoded_string.size() % 4)
    {
        encoded_string.push_back('='); // Padding to make the length a multiple of 4
    }

    return encoded_string;
}

vector<char> embed_message(const vector<char>& data, const string& message, int offset) {
    string base64_message = base64_encode(message);
    write_line("Base64-encoded message: " + base64_message);
    write_line("Base64 length (in characters): " + to_string(base64_message.length()));

    string binary_message;
    for (char ch : base64_message) {
        for (int i = 7; i >= 0; --i) {
            binary_message += ((ch >> i) & 1) ? '1' : '0';
        }
    }

    string length_bits;
    int length = base64_message.length();
    for (int i = 31; i >= 0; --i) {
        length_bits += ((length >> i) & 1) ? '1' : '0';
    }

    vector<char> modified_data = data;
    size_t i = 0;
    for (const char& bit : length_bits + binary_message) {
        modified_data[offset + i] = (modified_data[offset + i] & 0xFE) | (bit - '0');
        i++;
    }

    return modified_data;
}

void save_to_file(const vector<char>& data, const string& file_path) {
    FILE *output_file = fopen(file_path.c_str(), "wb");
    fwrite(data.data(), 1, data.size(), output_file);
    fclose(output_file);
}

int main() {
    string path = "/home/breezy/Documents/GitHub/Small-Projects/Thoth-Tech-Code-Files/steganography";
    string input_file_path = path + "/image.bmp";
    string output_file_path = path + "/encoded.bmp";
    
    FILE *input_file = fopen(input_file_path.c_str(), "rb");
    if (!input_file) {
        write_line("Error opening input file.");
        return 1;
    }

    fseek(input_file, 0, SEEK_END);
    size_t file_size = ftell(input_file);
    fseek(input_file, 0, SEEK_SET);

    vector<char> data(file_size);
    fread(data.data(), 1, file_size, input_file);
    fclose(input_file);

    write("Enter the message to embed: ");
    string message_to_embed = read_line();

    int pixel_data_offset = *reinterpret_cast<int*>(&data[10]);

    vector<char> encoded_data = embed_message(data, message_to_embed, pixel_data_offset);

    save_to_file(encoded_data, output_file_path);

    write_line("Message embedded successfully in Base64! Encoded image saved as " + output_file_path + ".");

    return 0;
}

File: steganography/programs/encode.py
import os
path = "/home/breezy/Documents/GitHub/Small-Projects/Thoth-Tech-Code-Files/steganography/programs/"

def base64_encode(input_string):
    BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    encoded_string = ""
    value = 0
    bits = -6
    base64_mask = 0x3F

    for character in input_string.encode():
        value = (value << 8) + character
        bits += 8

        while bits >= 0:
            encoded_string += BASE64_CHARS[(value >> bits) & base64_mask]
            bits -= 6
    
    if bits > -6:
        encoded_string += BASE64_CHARS[((value << 8) >> (bits + 8)) & base64_mask]
    
    while len(encoded_string) % 4:
        encoded_string += "="
    
    return encoded_string

def embed_message(data, message, offset):
    # Encode the message using Base64
    base64_message = base64_encode(message)
    print(f"Base64-encoded message: {base64_message}")
    print(f"Base64 length (in characters): {len(base64_message)}")

    binary_message = ''.join(format(ord(char), '08b') for char in base64_message)
    print(f"Binary message: {binary_message}")

    # Embed the length of the Base64 message (in bytes) in the first 32 bits
    length_bits = format(len(base64_message), '032b')
    print(f"Length bits: {length_bits}")

    data = bytearray(data)
    # Extract original binary at embedding positions
    original_binary = "".join(format(data[offset + i], '08b') for i in range(len(length_bits + binary_message)))
    modified_data = bytearray(data)

    # Embed the message into the least significant bits
    for i, bit in enumerate(length_bits + binary_message):
        modified_data[offset + i] = (modified_data[offset + i] & 0xFE) | int(bit)

    # Extract modified binary at embedding positions
    modified_binary = "".join(format(modified_data[offset + i], '08b') for i in range(len(length_bits + binary_message)))

    # Truncate for readability
    def truncate_binary(binary_str, show_bits=64):
        if len(binary_str) > show_bits * 2:
            return binary_str[:show_bits] + "..." + binary_str[-show_bits:]
        return binary_str

    print("\nOriginal binary data at embedding positions (truncated):")
    print(truncate_binary(original_binary))

    print("\nModified binary data at embedding positions (truncated):")
    print(truncate_binary(modified_binary))

    return modified_data


input_file_path = os.path.join(path, "cat.bmp")
output_file_path = os.path.join(path, "hiddencat.bmp")
with open(input_file_path, "rb") as f:data = f.read()

print("Enter the message to embed: ")
message_to_embed = input()

# Specify the offset where the pixel data starts (e.g., 54 for standard BMP)
pixel_data_offset = int.from_bytes(data[10:14], byteorder='little')

encoded_data = embed_message(data, message_to_embed, pixel_data_offset)
with open(output_file_path, "wb") as f: f.write(encoded_data)
print(f"Message embedded successfully in Base64! Encoded image saved as {output_file_path}.")
